src/base/utils/tar.cpp

```diff
@@ -0,0 +1,230 @@
+        // twice through the buffer.
+        void verifyAndCheckNull()
+        {
+            int calculatedChecksum = 0;
+            for (int i = 0; i < TAR_CHECKSUM_OFFSET; ++i)
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            calculatedChecksum += TAR_CHECKSUM_SIZE * ' '; // 8 ASCII spaces for the checksum field
+
+            for (int i = TAR_FILETYPE_OFFSET; i < m_header.size(); ++i)
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            const int checkSum = octalToDec(Utils::ByteArray::midView(m_header, TAR_CHECKSUM_OFFSET, TAR_CHECKSUM_SIZE));
+            m_null = (calculatedChecksum == TAR_CHECKSUM_SIZE * ' ') && (checkSum == 0);
+            m_verified = calculatedChecksum == checkSum;
+        }
+
+        bool m_null = true;
+        bool m_valid = false;
+        bool m_verified = false;
+
+        const QByteArray m_header;
+    };
+}
+
+
+QHash<QString, QByteArray> Utils::Tar::decompress(const QByteArray &data, bool *ok)
+{
+    if (ok)
+        *ok = false;
+    // Valid data data to us are at 1 header + payload
+    if (data.size() < TAR_HEADER_SIZE + 1)
+        return {};
+
+    int curPos = 0;
+    int dataSize = data.size();
+    QHash<QString, QByteArray> files;
+
+    while ((curPos < dataSize) && files.size() < 11) {
+        try {
+            TarEntry file(Utils::ByteArray::midView(data, curPos, TAR_HEADER_SIZE));
+            if (file.isNull()) // Possibly EOF
+                break;
+
+            if (!file.isValid()) {
+                LogMsg(QCoreApplication::translate("Utils::Tar",
+                                                   "Error while decompressing the GeoIP db. It has an invalid tar header.",
+                                                   "'tar' should be left untranslated. It is a filetype.")
+                       , Log::WARNING);
+                break;
+            }
+
+            if (!file.isVerified()) {
+                LogMsg(QCoreApplication::translate("Utils::Tar",
+                                                   "Error while decompressing the GeoIP db. A tar header checksum couldn't be verified.",
+                                                   "'tar' should be left untranslated. It is a filetype.")
+                       , Log::WARNING);
+                break;
+            }
+
+            const int fileSize = file.size();
+            const int padding = (fileSize > 0) ? (TAR_HEADER_SIZE - (fileSize % TAR_HEADER_SIZE))
+                                          : 0;
+
+            curPos += TAR_HEADER_SIZE + fileSize + padding;
+
+            if (!(file.type() == TarEntry::REG_FILE_ASCII || file.type() == TarEntry::REG_FILE_NULL))
+                continue;
+
+            const int filePos = curPos - fileSize - padding;
+            files.insert(file.name(), {data.constData() + filePos, fileSize});
```

Comment:
Should I construct the QByteArray here using fromRawData() for extra efficiency?

Reply:
You can't (shouldn't) do it! Then returned data is valid only until input data isn't disposed.